actors/ppo_atari_pz.py

```diff
@@ -512,22 +570,76 @@ def normalize_rewards(self, rewards: list, model: PPOModel) -> list:
         return normalized_reward
 
     def compute_gae(
-        self, rewards: torch.Tensor, values: torch.Tensor, dones: torch.Tensor, gamma: float = 0.99, lam: float = 0.95
-    ) -> torch.Tensor:
-        """Compute Generalized Advantage Estimation. See equations 11 & 12 in
+        self,
+        rewards: torch.Tensor,
+        values: torch.Tensor,
+        dones: torch.Tensor,
+        next_value: torch.Tensor,
+        gamma: float = 0.99,
+        lambda_: float = 0.95,
+    ):
+        """Compute Generalized Advantage Estimation (GAE). See equations 11 & 12 in
         https://arxiv.org/pdf/1707.06347.pdf
         """
-
         advs = []
         with torch.no_grad():
             gae = 0.0
-            # dones = torch.cat((dones, torch.zeros(1, 1).to(self._device)), dim=0)
+            # dones = torch.cat((dones, torch.zeros(1, 1).to(self._torch_device)), dim=0)
             for i in reversed(range(len(rewards))):
-                delta = rewards[i] + gamma * values[i + 1] * (1 - dones[i + 1]) - values[i]
-                gae = delta + gamma * lam * (1 - dones[i + 1]) * gae
+                delta = rewards[i] + gamma * next_value * (1 - dones[i]) - values[i]
+                gae = delta + gamma * lambda_ * (1 - dones[i]) * gae
                 advs.append(gae)
+                next_value = values[i]
         advs.reverse()
-        return torch.vstack(advs)
+
+        return advs
+
+
+class RolloutBuffer:
```

Comment:
It needs to be separate because `replay_buffer` is fundamentally differently than rollout buffer. Replay buffer means that a sample in this buffer can be reused more than 1 while a sample in rollout buffer can only be used once. In addition, rollout buffer does not have `sample` method and `self.next_observation`. So, I decided to separate them to avoid the confusion